tapfeatures/aux_channel_negotiator.go

```diff
@@ -0,0 +1,165 @@
+	"github.com/lightningnetwork/lnd/tlv"
+)
+
+// AuxChannelNegotiator is responsible for producing the extra tlv blob that is
+// encapsulated in the init and reestablish peer messages. This helps us
+// communicate custom feature bits with our peer.
+type AuxChannelNegotiator struct {
+	// peerFeatures keeps track of the supported features of our peers. This
+	// map will be used for lookups by other subsystems, when some features
+	// need to be supported by both parties to take effect.
+	peerFeatures lnutils.SyncMap[route.Vertex, *lnwire.RawFeatureVector]
+
+	// chanFeatures keeps track of the supported features of each channel.
+	// This map will be used for lookups by other subsystems to check
+	// whether certain custom channel features are supported.
+	chanFeatures lnutils.SyncMap[wire.OutPoint, *lnwire.RawFeatureVector]
+}
+
+// NewAuxChannelNegotiator returns a new instance of the aux channel negotiator.
+func NewAuxChannelNegotiator() *AuxChannelNegotiator {
+	return &AuxChannelNegotiator{}
+}
+
+// GetInitFeatures is called when sending an init message to a peer. It returns
+// custom feature bits to include in the init message TLVs. The implementation
+// can decide which features to advertise based on the peer's identity.
+func (n *AuxChannelNegotiator) GetInitFeatures(
+	peer route.Vertex) (tlv.Blob, error) {
+
+	var buf bytes.Buffer
+
+	// Grab the "static" feature vector that denotes the supported features
+	// of our node. If our peer can read this message they will keep track
+	// of our features just like we do below in `ProcessInitFeatures`.
+	features := getLocalFeatureVec()
+	err := features.Encode(&buf)
+	if err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
+}
+
+// ProcessInitFeatures handles received init feature TLVs from a peer. The
+// implementation can store state internally to affect future channel operations
+// with this peer.
+func (n *AuxChannelNegotiator) ProcessInitFeatures(peer route.Vertex,
+	features tlv.Blob) error {
+
+	buf := bytes.NewBuffer(features)
+	peerVec := lnwire.NewRawFeatureVector()
+	err := peerVec.Decode(buf)
+	if err != nil {
+		return err
+	}
+
+	// Before we store this peer's supported features we need to first check
+	// if our required features are supported by that peer. If a locally
+	// required feature is not supported by the remote peer we have to
+	// return an error and drop the connection. Whether we support all of
+	// the remote required features is a responsibility of the remote peer.
+	// If we fail to support a remotely required feature they are the ones
+	// to drop the connection (by returning an error right here).
+	err = checkRequiredBits(getLocalFeatureVec(), peerVec)
+	if err != nil {
+		return err
+	}
+
+	// Store this peer's features.
+	n.peerFeatures.Store(peer, peerVec)
+
+	return nil
+}
+
+// GetReestablishFeatures is called when sending a channel_reestablish message.
+// It returns feature bits based on the specific channel identified by its
```

Comment:
These arguments are unused similarly (though are of course unnamed here). Maybe more explicitly justify them in the comment if it's decided to keep them around.